Add d3d11va hardware video decoding

kevmitch opened the pull request:
This adds `hwdec=d3d11va-copy` using ffmpeg's d3d11va api. I'm not certain how or if this integrates with the parallel work by @wm4 on MediaFoundation decoder, but I'm issuing the pull request anyway so we can figure out how to move forward. 

The original intention was to reduce duplicated code between dxva2 and d3d11va. Therefore all of the shared code has been put into a separate decode/d3d.c file. I've stopped short of rewriting dxva2.c to use it though to see how things stand first with the MediaFoundation stuff.

wm4 commented on ta/ta_talloc.h:
```diff
@@ -57,6 +57,8 @@
 #define talloc_asprintf_append          ta_talloc_asprintf_append
 #define talloc_asprintf_append_buffer   ta_talloc_asprintf_append_buffer
 
+#define TALLOC_FREE(ctx) do {talloc_free(ctx); ctx=NULL;} while(0)
```
Better quote ctx as `(ctx)=NULL;`.

wm4 commented:
>The original intention was to reduce duplicated code between dxva2 and d3d11va. Therefore all of the shared code has been put into a separate decode/d3d.c file. I've stopped short of rewriting dxva2.c to use it though to see how things stand first with the MediaFoundation stuff.

I think that should be fine. The things that can possibly be shared with MF are device creation and VO interop. Everything else is taken care of MF itself.

Also, maybe it should be named TALLOC_FREEP() or so. (ffmpeg calls its equivalent av_freep, while av_free is the normal free.)

kevmitch commented:
> Also, maybe it should be named TALLOC_FREEP() 

even though this contradicts the upstream talloc api?

wm4 commented:
If that is so, maybe we should go with a TA_FREEP()?

kevmitch commented:
Just realized that

```c
void av_freep(void **ptr_to_free)
```

but if the macro were a function, it's declaration would be 

```c
void TA_FREEP(void *ptr_to_free)
```

Should I just make a proper function 
```c
void ta_freep(void **ptr_to_free)
```
to match ffmpeg?

wm4 commented:
No, I dislike this aliasing nonsense ffmpeg has to go through to even make av_freep valid C. It also leads to mistakes easily by accidentally passing a pointer instead of pointer-to-pointer.

wm4 commented:
To clarify, it's 
```C
void av_freep(void *);
```

kevmitch commented:
So just live with the inconsistency?:

```c 
av_freep(&s->avctx->hwaccel_context);
TA_FREEP(s->hwdec_ctx);
```

wm4 commented:
Oh, I guess now I get it. We can just pass the pointer to the macro too, no harm in that. Actually it should be safer.

kevmitch commented:
* TA_FREEP is now as discussed.
* moved `ID3D10Multithread_SetMultithreadProtected(multithread, FALSE)` into block for setting up direct3d11 device for use only within d3d11va.c. As discussed on IRC, when using an external device (from the vo), multithredprotected must be TRUE.
* fixed leak, where i didn't free the `p->decoder` structure in `video/decode/d3d11va.c:d3d11va_uninit`
* fixed release of incorrect object in `video/d3d11va.c:d3d11va_release_img` 
* remove `p->decoder->cfg` from `video/decode/d3d11va` and just make it the talloc child of `p->decoder` before passing it to ffmpeg in `video/decode/d3d11va:d3d11va_init_decoder` . It is not needed anywhere else.
* changed `d3d_decoder_guid_to_name` to `d3d_decoder_guid_to_desc_buf` so that it moves the GUID to string code into the shared `video/decode/d3d.c` file.
* rewrite `video/decode/dxva2.c` to use the shared code
* merge `video/d3d.h` into `video/dxva2.h` and rename `mp_d3d_ctx` to `mp_dxva2_ctx` to clarify nomenclature. 
* Stop using `LPDIRECT3DSURFACE9`

kevmitch commented:
Seems the travis error is genuine. libav stable doesn't define `AV_PIX_FMT_D3D11VA_VLD`. Should I add a configure check?

kevmitch commented:
wait d3d11va isn't in libav stable at all. I should probably just put guards around it's use in 
`#ifdef HAVE_D3D11VA` guards around its use in `fmt-conversions.c`

kevmitch commented:
fixed

kevmitch commented:
@wm4 or would it work better with MF to keep a separate `video/d3d.h` and keep the structure as `mp_d3d_ctx` adding a ID3D11Device pointer such that only one device is actually set?

```c
struct mp_d3d_ctx {
    struct mp_hwdec_ctx hwctx;
    IDirect3DDevice9 *d3d9_device;
    ID3D11Device     *d3d11_device;
};

wm4 commented on video/d3d11va.c:
```diff
@@ -0,0 +1,86 @@
+    struct mp_image *mpi = mp_image_new_custom_ref(
+        &(struct mp_image){0}, surface, d3d11va_release_img);
+    if (!mpi)
+        return NULL;
```
In theory leaks a lot of things in the error case, because mp_image_new_custom_ref doesn't unref anything on failure. Since this fails only if very small memory allocations fail, you could replace this with an abort().

wm4 commented on video/d3d11va.c:
```diff
@@ -0,0 +1,86 @@
+        return NULL;
+    struct d3d11va_surface *surface = talloc_zero(NULL, struct d3d11va_surface);
+
+    surface->d3d11_dll = LoadLibrary(L"d3d11.dll");
```
Can't we assume that the library must always have been loaded at this point, and just call GetModuleHandle?

Or maybe even define a global HANDLE for this which would get lazy-loaded with pthread_init_once().

kevmitch replied:
Yes we can, but we can't assume that it will stay loaded for the lifetime of the mp_image. At least in the case of `d3d9.dll` not having this caused a crash with dxva2 when the library was freed before the image. The library handle itself isn't actually needed here. It is just here to increment the reference count on the library.

> The GetModuleHandle function returns a handle to a mapped module without incrementing its reference count.

kevmitch replied:
So presumably, the global handle solution would never call FreeLibrary? Where would the global handle live? It doesn't look like anything else is currently using pthread_init_once to model off of.

wm4 replied:
Yes. there's no reason to unload it after all. We only dynamically load it because it might be unavailable on the system. Anyway, this is just a suggestion and orthogonal to this PR.

wm4 commented on video/decode/d3d11va.c:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    // TODO: VO supplied device
+    {
```
Maybe move this code block into a function? That's going to happen anyway.

wm4 commented on video/decode/d3d11va.c:
```diff
@@ -0,0 +1,484 @@
+            MP_ERR(p, "Failed to get Multithread interface: %s\n",
+                   mp_HRESULT_to_str(hr));
+        }
+        ID3D10Multithread_SetMultithreadProtected(multithread, FALSE);
```
Note that the default is FALSE.

wm4 commented on video/decode/vd_lavc.c:
```diff
@@ -143,6 +144,9 @@ static const struct vd_lavc_hwdec *const hwdec_list[] = {
 #endif
+#if HAVE_D3D11VA_HWACCEL
+    &mp_vd_lavc_d3d11va_copy,
+#endif
```
Is it really ok to place it here? This will prefer d3d11 over d3d9, so copying is preferred over directed GL or ANGLE interop.

kevmitch replied:
What should the ultimate order be?

d3d11va
dxva2
d3d11va-copy
dxva2-copy

? The downside is it interleaves the two requiring duplicated macro guards. But the alternative is that `dxva2-copy` gets preferred over `d3d11va` once that's added.

kevmitch replied:
also is there a reason why the `enum hwdec_type` in `video/hwdec.h` is explicitly numbered?

wm4 commented:
Made some comments, but overall looks good I guess.

>@wm4 or would it work better with MF to keep a separate video/d3d.h and keep the structure as mp_d3d_ctx adding a ID3D11Device pointer such that only one device is actually set?

Probably. I mean, the D3D11 device also will need to be passed along, right? I'm not quite sure why you renamed the struct.

kevmitch commented:
> Probably. I mean, the D3D11 device also will need to be passed along, right? I'm not quite sure why you renamed the struct.

I guess I was originally thinking that d3d11 would get its own struct to keep them separate and eliminate the need for the extra generic `d3d.h` file, but given that it sounds like MF can use either d3d9 or d3d11, i guess it is better to put them together.

wm4 commented:
I guess I agree.

kevmitch commented:
Maybe I should put the `video/decode/d3d.c` one directory up in `video`, add `struct mp_d3d_ctx` as well as functions to create d3d9 or d3d11 devices.

kevmitch commented:
Pushed changes:
* `assert(0)` on `mp_image_new_custom_ref` failure  for both `d3d11va` and `dxva2`
* moved device creation into local functions for both `d3d11va` and `dxva2`. for d3d11va there is a `thread_safe` argument that is currently set to `false`  for `d3d11va-copy`.
* reorder d3d hwdecs to interleaved order
* revert change to `mp_d3d_ctx`.

Further planning:
* Global `d3d` library init
* As described above, we could move `video/decode/d3d.c` down to `video/d3d.c` with `struct mp_d3d_ctx`, add generic device creation functions (aided by global library init). The VOs  could then also use these device creation functions.
* Would it make sense to put the `dxva2_new_ref` and `d3d11va_new_ref` in there too?

wm4 commented:
> also is there a reason why the `enum hwdec_type` in `video/hwdec.h` is explicitly numbered? 

I think no. Will double-check.

>  assert(0) on mp_image_new_custom_ref failure for both d3d11va and dxva2 

It should be abort(), because it's _not_ unreachable code, just an absurdly unlikely actual failure.

kevmitch commented:
>It should be abort(), because it's not unreachable code, just an absurdly unlikely actual failure.

Fixed. Was not aware of that convention.

kevmitch commented:
> Was not aware of that convention.

Although it makes sense the more I think about it.

kevmitch commented:
Switched probing order so that d3d11va is last. Added a blurb to the man page that it is experimental. Should I push?

kevmitch commented:
use FALSE literal for BOOL

wm4 commented:
>Should I push?

Sure.
